In [1]:
import pandas as pd
import numpy as np
import csv
import os
import copy
import json

In [2]:
train_path_ = '../data/train.txt'
test_path = '../data/dev.txt'
label_path = '../data/label.txt'

In [3]:
def data_loader(path):
    with open(path, "r", encoding="utf-8") as train:
        reader = csv.reader(train, delimiter="\t")
        text_file = []
        for line in reader:
            text_file.append(line)
    return text_file

#Converting upload file into file with binary classification, where 1 - ISA relation, 0 - other relation

def binary_convert(path):
    txt_file = data_loader(path)
    text_array = np.array(txt_file)

    copy_arr = text_array
    for row_idx in range(len(text_array)):
        if text_array[row_idx][1] == 'ISA':
              copy_arr[row_idx][1] = 1
        elif text_array[row_idx][1] != 'ISA':
              copy_arr[row_idx][1] = 0
    print(f'Length of uploaded file:{len(copy_arr)}')
    return copy_arr

#We create two arrays ISA_example and Other_example for dividing the train file on train and evaluation sets

def create_new_arrays(path):
    binary_text = binary_convert(path)
    ISA_example = []
    Other_example = []
    for line_idx in range(len(binary_text)):
        if binary_text[line_idx][1] == '1':
              ISA_example.append(binary_text[line_idx])
        else:
              Other_example.append(binary_text[line_idx])
    print(f'List with ISA relations was created, Lenght: {len(ISA_example)}')
    print(f'List with Other relations was created, Lenght: {len(Other_example)}')
    return ISA_example, Other_example

def convert_in_dataframe(file):
    df = pd.DataFrame(columns = ['text', 'label'])
    text = []
    label = []
    for idx in range(len(file)):
        text.append(file[idx][0])
        label.append(file[idx][1])

    df['text'] = text
    df['label'] = label
    return df

In [7]:
train_ISA, train_Other = create_new_arrays(train_path_)

Length of uploaded file:659
List with ISA relations was created, Lenght: 94
List with Other relations was created, Lenght: 565


We divide test_file into two lists: test_text includes only texts, test_labels includes only labels for each text from test_textt. It is nessesary for prediction and getting metrics.

In [8]:
test_file = binary_convert(test_path)

Length of uploaded file:74


In [9]:
test_text = []
test_label = []

for text_raw,label in test_file:
    test_text.append(text_raw)
    test_label.append(label)
assert len(test_text) == len(test_label)

In [7]:
test_text_df = pd.DataFrame(test_text)
test_label_df = pd.DataFrame(test_label)

In [19]:
test_text_df.to_csv('data/test_text.csv', index = False, header=False, sep = '\t')
test_label_df.to_csv('data/test_label.csv', index = False, header=False, sep = '\t')

We divide our train file on train and evaluation sets, where train set includes 65 texts with ISA relations and 200 texts with other relations, evaluation set includes 30 texts with ISA relations and 150 texts with other relations

In [6]:
train_set = []
for idx in range(len(train_ISA)):
    if idx < 65:
        train_set.append(train_ISA[idx])

for idx in range(len(train_Other)):
    if idx < 200:
        train_set.append(train_Other[idx])

np.random.shuffle(train_set)

eval_set = []
for idx in range(len(train_ISA)):
    if idx > 64:
        eval_set.append(train_ISA[idx])

for idx in range(len(train_Other)):
    if idx > 199 and idx < 350:
        eval_set.append(train_Other[idx])

np.random.shuffle(eval_set)

Save new txt files with binary classification in csv format

In [7]:
train_df = convert_in_dataframe(train_set)
train_df.to_csv('data/train_file.csv', index = False, header=False, sep = '\t')

In [8]:
eval_df = convert_in_dataframe(eval_set)
eval_df.to_csv('data/eval_file.csv', index = False, header=False, sep = '\t')

In [9]:
test_df = convert_in_dataframe(test_file)
test_df.to_csv('data/test_file.csv', index = False, header=False, sep = '\t')

Create train set and eval with balanced classes

Length of uploaded file: 659


List with ISA relations was created, Lenght: 94


List with Other relations was created, Lenght: 565


Select the 70 ISA rlations and 70 Other relations for train
And 24 ISA rlations and 100 Other relations for eval

In [28]:
train_balanced = []
for idx in range(len(train_ISA)):
    if idx < 70:
        train_balanced.append(train_ISA[idx])

for idx in range(len(train_Other)):
    if idx < 70:
        train_balanced.append(train_Other[idx])

np.random.shuffle(train_balanced)

eval_balanced = []
for idx in range(len(train_ISA)):
    if idx > 69:
        eval_balanced.append(train_ISA[idx])

for idx in range(len(train_Other)):
    if idx > 69 and idx < 169:
        eval_balanced.append(train_Other[idx])

np.random.shuffle(eval_balanced)

In [30]:
train_balanced_df = convert_in_dataframe(train_balanced)
train_balanced_df.to_csv('data/train_balanced.csv', index = False, header=False, sep = '\t')

In [31]:
eval_balanced_df = convert_in_dataframe(eval_balanced)
eval_balanced_df.to_csv('data/eval_balanced.csv', index = False, header=False, sep = '\t')

Create train set and eval with balanced classes
Select the 70 ISA rlations and 350 Other relations for train And 24 ISA rlations and 200 Other relations for eval

In [32]:
train_unbalanced = []
for idx in range(len(train_ISA)):
    if idx < 70:
        train_unbalanced.append(train_ISA[idx])

for idx in range(len(train_Other)):
    if idx < 350:
        train_unbalanced.append(train_Other[idx])

np.random.shuffle(train_unbalanced)

eval_unbalanced = []
for idx in range(len(train_ISA)):
    if idx > 69:
        eval_unbalanced.append(train_ISA[idx])

for idx in range(len(train_Other)):
    if idx > 349 and idx < 550:
        eval_unbalanced.append(train_Other[idx])

np.random.shuffle(eval_unbalanced)

In [33]:
train_unbalanced_df = convert_in_dataframe(train_unbalanced)
train_unbalanced_df.to_csv('data/train_unbalanced.csv', index = False, header=False, sep = '\t')

In [34]:
eval_unbalanced_df = convert_in_dataframe(eval_unbalanced)
eval_unbalanced_df.to_csv('data/eval_unbalanced.csv', index = False, header=False, sep = '\t')